### PR TITLE
feat: Add plugin-kit package

### LINKED ISSUE #97: Change Request: New "helper" package for plugins
### Which packages would you like to change?

- [ ] `@eslint/compat`
- [ ] `@eslint/config-array`
- [ ] `@eslint/core`
- [ ] `@eslint/migrate-config`
- [ ] `@eslint/object-schema`

### What problem do you want to solve?

Now that I've worked on a couple of language plugins, it's clear that there's some duplication of logic that each language could have. For instance:

- A basic `SourceCode` implementation
- Classes for `VisitTraversalStep` and `CallTraversalStep`
- Configuration comment parsing

### What do you think is the correct solution?

Rather than forcing everyone to write their own versions of this logic, it would be helpful to create a package that contains common utilities for language development.

Maybe `@eslint/language`? Or `@eslint/plugin` or `@eslint/plugin-kit` (to disambiguate with ESLint plugins and make it a more generic plugin helper package rather than a language plugin package?)

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

-   [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Created the new `plugin-kit` package

#### What changes did you make? (Give an overview)

- Created `plugin-kit` directory
- Created `ConfigCommentParser`, `VisitNodeStep`, and `CallMethodStep` classes
- Added tests for each class
- Updated release-please config
- Updated release-please CI job
- Updated issue templates

#### Related Issues

fixes #97

#### Is there anything you'd like reviewers to focus on?

Just starting with the basics for this package. These are some commonalities I've found across the different languages. There will likely be more.

I slightly changed the interface of `ConfigCommentParser` to be more generic as the one in `eslint/eslint` was very specific to the way Espree represents comments. I tested this `ConfigCommentParser` in ESLint directly to ensure the new interface works correctly.

### CODE REVIEW

**Review by harish-sethuraman — APPROVED** [review]
LGTM. Went through the files related to actions/readme/build. Will wait for others to review files in source and test folders

File: packages/plugin-kit/src/config-comment-parser.js [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+			 * For example, the following configuration is invalid:
+			 *
+			 *    "no-alert: 2 no-console: 2"
+			 *
+			 * This results in a configuration of { "no-alert": "2 no-console: 2" }, which is
+			 * not valid. In this case, the configuration should be ignored.
+			 */
+			if (isEverySeverityValid(items)) {
+				return {
+					ok: true,
```
[the remark this answers is not in the document]
Comment (nzakas) [reply packages/plugin-kit/src/config-comment-parser.js:142]: Yes that's correct.

File: packages/plugin-kit/src/config-comment-parser.js [hunk 2]
```diff
@@ -0,0 +1,244 @@
+/**
+ * @fileoverview Config Comment Parser
+ * @author Nicholas C. Zakas
+ */
+
+/* eslint class-methods-use-this: off -- Methods desired on instance */
```
[the remark this answers is not in the document]
Comment (nzakas) [reply packages/plugin-kit/src/config-comment-parser.js:6]: Making it static class means we'll never be able to pass parameters in, and I think there's a high likelihood that we will do so in the future.

File: packages/plugin-kit/jsr.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "@eslint/plugin-kit",
+  "version": "0.0.0",
+  "exports": "./dist/esm/index.js",
+  "publish": {
+    "include": [
+      "dist/esm/index.js",
+      "dist/esm/index.d.ts",
+      "dist/esm/types.ts",
+      "dist/esm/types.d.ts",
```
[the remark this answers is not in the document]
Comment (nzakas) [reply packages/plugin-kit/jsr.json:10]: I'm following the convention we have in the other packages. If we want to change that convention, let's do that for all packages in a separate PR.

**Review by aladdin-add — APPROVED** [review]
LGTM. Would like another review before merging.

File: packages/core/src/types.ts [hunk 4]
```diff
@@ -402,7 +402,7 @@ export type SourceCode = TextSourceCode | BinarySourceCode;
  */
 export interface VisitTraversalStep {
 	kind: 1;
-	target: object;
+	target: unknown;
```
Comment (mdjermanovic) [inline packages/core/src/types.ts:405]: Why this change, isn't this expected to be an object (AST node)?

File: release-please-config.json [hunk 5]
```diff
@@ -45,6 +45,17 @@
     },
     "packages/migrate-config": {
       "release-type": "node"
+    },
+    "packages/plugin-kit": {
+      "release-type": "node",
+      "release-as": "0.1.0",
```
Comment (mdjermanovic) [inline release-please-config.json:51]:
```suggestion
```

I think this isn't necessary, and we'd need to remove it after the first release so seems better not to add it.

Comment (mdjermanovic) [reply release-please-config.json:51]: Actually it does seem necessary if we want to start with 0.x, I see now that new packages were released as 1.0.0.

File: packages/plugin-kit/README.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+// create a new instance
+const commentParser = new ConfigCommentParser();
+
+// list format
+const list = commentParser.parseListConfig("prefer-const, semi");
+console.log(Object.entries(list)); // [["prefer-const", true], ["semi", true]]
+
+// string format
+const strings = commentParser.parseStringConfig("foo:off, bar");
+console.log(Object.entries(list)); // [["foo", "off"], ["bar", null]]
```
Comment (mdjermanovic) [inline packages/plugin-kit/README.md:72]:
```suggestion
console.log(Object.entries(strings)); // [["foo", "off"], ["bar", null]]
```

File: packages/plugin-kit/README.md [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+// JSON-like config format
+const jsonLike = commentParser.parseJSONLikeConfig(
+	"semi:[error, never], prefer-const: warn",
+);
+console.log(Object.entries(list)); // [["semi", ["error", "never"]], ["prefer-const", "warn"]]
```
Comment (mdjermanovic) [inline packages/plugin-kit/README.md:78]:
```suggestion
console.log(Object.entries(jsonLike.config)); // [["semi", ["error", "never"]], ["prefer-const", "warn"]]
```

File: packages/core/src/types.ts (hunk 4 above)
Comment (nzakas) [reply packages/core/src/types.ts:405]: `unknown` forces consumers to specify the type of the object before working with it, which ensures that it's properly typed.

`object` will just cause squiggles every time you access a property or method that isn't on `Object.prototype`.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**